src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs

```diff
@@ -176,40 +176,32 @@ public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellati
 
             try
             {
-                // This variable records any error occurring in the thread pool task's context.
-                Exception error = null;
-
-                await Task.Run(() =>
-                {
-                    try
-                    {
-                        this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).Wait(cancellation);
-                    }
-                    catch (Exception e)
-                    {
-                        // Record the error occurring in the thread pool's context.
-                        error = e;
-                    }
-                }).ConfigureAwait(false);
-
-                // Throw the error within this error handling context.
-                if (error != null)
-                    throw error;
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttempt(false, endPoint));
+                await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).WithCancellationAsync(cancellation).ConfigureAwait(false);
+                this.asyncProvider.Signals.Publish(new PeerConnected(false, endPoint));
 
                 this.stream = this.tcpClient.GetStream();
             }
             catch (OperationCanceledException)
             {
-                this.logger.LogDebug("Connecting to '{0}' cancelled.", endPoint);
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttemptFailed(false, endPoint, "Operation Canceled"));
+                this.logger.LogTrace("Connecting to '{0}' cancelled.", endPoint);
                 this.logger.LogTrace("(-)[CANCELLED]");
                 throw;
             }
-            catch (Exception e)
+            catch (SocketException ex)
+            {
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttemptFailed(false, endPoint, $"Socket Exception: {ex.Message}"));
+                this.logger.LogDebug("Error connecting to '{0}', exception message: {1}", endPoint, ex.Message);
+                this.logger.LogTrace("(-)[UNHANDLED_EXCEPTION]");
+                throw;
+            }
+            catch (Exception ex)
             {
-                if (e is AggregateException) e = e.InnerException;
```

Comment:
I am not sure .Message is enough in an aggregate exception.

Also you are changing the behaviour here, before it would throw the inner exception (just an FYI).

Reply:
forget to reply to this one.
The behavior is the same, just with other code:
`await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).WithCancellationAsync(cancellation).ConfigureAwait(false);` doesn't need to handle an aggregate exception because you need it only when you deal with plain tasks that doesn't throw exceptions, while using the async keyword exception can be catch directly

Reply:
I am not sure I follow, in anycase you aggregate exception can be thrown not only with TPL.
It might not be an issue though.